### PR TITLE
[azcore] Remove logger from azcore

### PR DESCRIPTION
**Pull request by seankane-msft**

### CODE REVIEW

File: sdk/azcore/example_test.go [hunk 1]
```diff
@@ -48,23 +48,23 @@ func ExampleRequest_SetBody() {
...
 
-func ExampleLogger_Should() {
-	// you can create your own logging classification as needed
-	const LogExpensiveThing azcore.LogClassification = "ExpensiveThing"
-	if azcore.Log().Should(LogExpensiveThing) {
-		// perform expensive calculation only when enabled
-		azcore.Log().Write(LogExpensiveThing, "expensive log message")
-	}
-}
+// func ExampleLogger_Should() {
```
Comment (chamons) [inline sdk/azcore/example_test.go:51]: Reason to comment out and not remove?

File: sdk/azcore/ci.yml [hunk 2]
```diff
@@ -3,6 +3,7 @@ trigger:
   paths:
     include:
     - sdk/azcore/
+    - eng/
```
Comment (chamons) [inline sdk/azcore/ci.yml:6]: How is this related?

File: sdk/azcore/policy_logging.go [hunk 3]
```diff
@@ -35,7 +36,7 @@ func NewLogPolicy(o *LogOptions) Policy {
 	return &logPolicy{options: *o}
 }
 
-// logPolicyOpValues is the struct containing the per-operation values
+// logPolicyOpValues is the strucsst containing the per-operation values
```
Comment (chamons) [inline sdk/azcore/policy_logging.go:39]: typo?

File: sdk/internal/ci.yml [hunk 4]
```diff
@@ -3,12 +3,13 @@ trigger:
   paths:
     include:
     - sdk/internal/
+    - eng/
 
 pr:
   paths:
     include:
     - sdk/internal/
-
```
Comment (chamons) [inline sdk/internal/ci.yml:11]: White space.

File: sdk/internal/ci.yml [hunk 5]
```diff
@@ -3,12 +3,13 @@ trigger:
   paths:
     include:
     - sdk/internal/
+    - eng/
```
Comment (chamons) [inline sdk/internal/ci.yml:6]: Same as above, why needed?

**Review by chamons — COMMENTED** [review]
Is this ready to land or should it be in draft?

File: sdk/internal/logger/logger.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+	"os"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+)
+
+// Listener is the function signature invoked when writing log entries.
+// A Listener is required to perform its own synchronization if it's
+// expected to be called from multiple Go routines.
+type Listener func(azcore.LogClassification, string)
```
Comment (catalinaperalta) [inline sdk/internal/logger/logger.go:17]: Copy the `azcore.LogClassification` type over to the internal module. Then in `azcore` type alias the `LogClassification` type over there if if you need it.

Comment (catalinaperalta) [reply sdk/internal/logger/logger.go:17]: All the below references to the `azcore.LogClassification` will then be to the one referenced here to remove the dependency on `azcore`.

File: sdk/azcore/log_test.go [hunk 7]
```diff
@@ -9,46 +9,48 @@ import (
...
 	"testing"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/internal/logger"
 )
 
 func TestLoggingDefault(t *testing.T) {
 	// ensure logging with nil listener doesn't fail
-	Log().SetListener(nil)
-	Log().Write(LogRequest, "this should work just fine")
+	SetListener(nil)
```
Comment (catalinaperalta) [inline sdk/azcore/log_test.go:18]: Wondering if these two test should be moved to the internal module?

Comment (catalinaperalta) [reply sdk/azcore/log_test.go:18]: In azcore, you could add tests for SetListener and SetClassification.

File: sdk/internal/logger/logger.go [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+	// This includes information like the HTTP status code, headers, and request URL.
+	LogResponse LogClassification = "Response"
+
+	// LogRetryPolicy entries contain information specific to the retry policy in use.
+	LogRetryPolicy LogClassification = "RetryPolicy"
+
+	// LogLongRunningOperation entries contain information specific to long-running operations.
+	// This includes information like polling location, operation state and sleep intervals.
+	LogLongRunningOperation LogClassification = "LongRunningOperation"
+)
```
Comment (catalinaperalta) [inline sdk/internal/logger/logger.go:30]: Just the type declaration needs to stay be in the internal module, these should be in `azcore` so that customers can use them.

File: sdk/internal/logger/logger_test.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+)
+
+func TestLoggingDefault(t *testing.T) {
+	// ensure logging with nil listener doesn't fail
+	azcore.SetListener(nil)
+	Log().Write(azcore.LogRequest, "this should work just fine")
+
+	log := map[azcore.LogClassification]string{}
+	azcore.SetListener(func(cls azcore.LogClassification, msg string) {
```
Comment (catalinaperalta) [inline sdk/internal/logger/logger_test.go:20]: These tests in `internal` are using `azcore`.

File: sdk/internal/logger/logger_test.go [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+		log[cls] = msg
+	})
+	const req = "this is a request"
+	Log().Write(azcore.LogRequest, req)
```
Comment (catalinaperalta) [inline sdk/internal/logger/logger_test.go:24]: Hmm...I think you can declare unexported versions of the log classifications to use in these tests...or you could leave the declarations you have here and again alias each one in azcore.

File: sdk/azcore/log.go [hunk 11]
```diff
@@ -31,86 +28,20 @@ const (
...
-		logger.lst = func(cls LogClassification, msg string) {
-			// simple console logger, it writes to stderr in the following format:
-			// [time-stamp] Classification: message
-			fmt.Fprintf(os.Stderr, "[%s] %s: %s\n", time.Now().Format(time.StampMicro), cls, msg)
-		}
-	}
+func SetListener(lst interface{}) {
+	l := lst.(logger.Listener)
+	logger.Log().SetListener(l)
 }
```
Comment (seankane-msft) [inline sdk/azcore/log.go:47]: @catalinaperalta the input to `SetListener` will have to be a `logger.Listener` type unless there is a way to "show" the function that LogClassification is the same both ways. What are your thoughts here?

File: sdk/azcore/log_test.go (hunk 7 above)
Comment (seankane-msft) [reply sdk/azcore/log_test.go:18]: I like having these tests here because they do a good job of making sure the external functionality hasn't changed drastically.

Comment (catalinaperalta) [reply sdk/azcore/log_test.go:18]: I see, ok makes sense.

File: sdk/azcore/go.sum [hunk 12]
```diff
@@ -1,15 +1,36 @@
 github.com/Azure/azure-sdk-for-go/sdk/internal v0.5.1 h1:vx8McI56N5oLSQu8xa+xdiE0fjQq8W8Zt49vHP8Rygw=
 github.com/Azure/azure-sdk-for-go/sdk/internal v0.5.1/go.mod h1:k4KbFSunV/+0hOHL1vyFaPsiYQ1Vmvy1TBpmtvCDLZM=
+github.com/Azure/azure-sdk-for-go/sdk/internal v0.5.2 h1:E2xwjsWU81O/XuSaxAGa8Jmqz4Vm4NmrpMSO9/XevDg=
```
Comment (catalinaperalta) [inline sdk/azcore/go.sum:3]: Run `go mod tidy`

File: sdk/azcore/log.go [hunk 13]
```diff
@@ -7,12 +7,11 @@ package azcore
...
 	"fmt"
-	"os"
-	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/internal/logger"
 )
 
-// LogClassification is used to group entries.  Each group can be toggled on or off.
-type LogClassification string
+type LogClassification logger.LogClassification
```
Comment (catalinaperalta) [inline sdk/azcore/log.go:14]: Keep the comment.

File: sdk/azcore/log.go [hunk 14]
```diff
@@ -31,86 +30,34 @@ const (
...
-// If no listener has been set this will return false.
-// Calling this method is useful when the message to log is computationally expensive
-// and you want to avoid the overhead if its log classification is not enabled.
-func (l *Logger) Should(cls LogClassification) bool {
-	if l.lst == nil {
-		return false
-	}
-	if l.cls == nil || len(l.cls) == 0 {
-		return true
+func SetClassifications(cls ...LogClassification) {
```
Comment (catalinaperalta) [inline sdk/azcore/log.go:33]: Also needs the comment.

File: sdk/azcore/log.go [hunk 15]
```diff
@@ -31,86 +30,34 @@ const (
...
 
-// Write invokes the underlying Listener with the specified classification and message.
-// If the classification shouldn't be logged or there is no listener then Write does nothing.
-func (l *Logger) Write(cls LogClassification, message string) {
-	if !l.Should(cls) {
-		return
-	}
-	l.lst(cls, message)
-}
+type Listener func(LogClassification, string)
```
Comment (catalinaperalta) [inline sdk/azcore/log.go:41]: Doc comment.

File: sdk/azcore/log.go [hunk 16]
```diff
@@ -31,86 +30,34 @@ const (
...
-}
-
-var logger Logger
-
-// Log returns the process-wide logger.
-func Log() *Logger {
-	return &logger
+func SetListener(lst Listener) {
+	if lst == nil {
+		fmt.Println("nil listener")
```
Comment (catalinaperalta) [inline sdk/azcore/log.go:52]: Did you want to log here?

Comment (catalinaperalta) [reply sdk/azcore/log.go:52]: Same for the below println

File: sdk/azcore/log.go [hunk 17]
```diff
@@ -31,86 +30,34 @@ const (
...
+		logger.Log().SetListener(nil)
+	} else {
+		fmt.Println("Not nil listener")
+		logger.Log().SetListener(transform(lst))
```
Comment (catalinaperalta) [inline sdk/azcore/log.go:56]: You can just put the code that's in transform() directly here in `SetListener()`

File: sdk/azcore/log_test.go [hunk 18]
```diff
@@ -9,21 +9,23 @@ import (
...
+	logger.Log().Write(logger.LogRequest, "this should work just fine")
```
Comment (catalinaperalta) [inline sdk/azcore/log_test.go:19]: Since this test is in azcore, should you be testing with `azcore.LogRequest`? Same for the other classifications in these tests.

File: sdk/azcore/poller.go [hunk 19]
```diff
@@ -200,13 +202,13 @@ func (l *LROPoller) FinalResponse(ctx context.Context, respType interface{}) (*h
 // into the respType interface that is provided.
 func (l *LROPoller) PollUntilDone(ctx context.Context, freq time.Duration, respType interface{}) (*http.Response, error) {
 	logPollUntilDoneExit := func(v interface{}) {
-		Log().Writef(LogLongRunningOperation, "END PollUntilDone() for %T: %v", l.lro, v)
+		logger.Log().Writef(logger.LogLongRunningOperation, "END PollUntilDone() for %T: %v", l.lro, v)
```
Comment (catalinaperalta) [inline sdk/azcore/poller.go:205]: I think these also should reference the `azcore` instances, just feels more correct. 
```suggestion
		logger.Log().Writef(azcore.LogLongRunningOperation, "END PollUntilDone() for %T: %v", l.lro, v)
```

File: sdk/azcore/log.go [hunk 20]
```diff
@@ -7,12 +7,11 @@ package azcore
...
 import (
 	"fmt"
-	"os"
-	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/internal/logger"
 )
 
-// LogClassification is used to group entries.  Each group can be toggled on or off.
-type LogClassification string
```
Comment (catalinaperalta) [inline sdk/azcore/log.go:15]: Commenting here for the consts below I think you should do the following since the values are the same as those declared in the `internal` module and it would be better to have the string value in one place for future modifications:
```
const (
	// LogRequest entries contain information about HTTP requests.
	// This includes information like the URL, query parameters, and headers.
	LogRequest LogClassification = logger.LogRequest 
	// LogResponse entries contain information about HTTP responses.
	// This includes information like the HTTP status code, headers, and request URL.
	LogResponse LogClassification = logger.LogResponse 
	// LogRetryPolicy entries contain information specific to the retry policy in use.
	LogRetryPolicy LogClassification = logger.LogRetryPolicy 
	// LogLongRunningOperation entries contain information specific to long-running operations.
	// This includes information like polling location, operation state and sleep intervals.
	LogLongRunningOperation LogClassification = logger.LogLongRunningOperation 
)
```

File: sdk/azcore/log_test.go (hunk 18 above)
Comment (seankane-msft) [reply sdk/azcore/log_test.go:19]: This calls into the logger module so it has to use the type definitions located in that module.

File: sdk/azcore/log.go (hunk 17 above)
Comment (seankane-msft) [reply sdk/azcore/log.go:56]: I added a doc comment for future Sean to understand what the transform does and why it is there, I'd like to keep it as a separate function for readability.

File: sdk/azcore/log.go (hunk 20 above)
Comment (seankane-msft) [reply sdk/azcore/log.go:15]: The compiler complains about this because the types don't match up (logger.LogClassification vs string types). I think we are stuck with having these definitions in two separate places.

File: sdk/azcore/poller.go (hunk 19 above)
Comment (seankane-msft) [reply sdk/azcore/poller.go:205]: Same comment as above.

File: sdk/azcore/log.go (hunk 20 above)
Comment (seankane-msft) [reply sdk/azcore/log.go:15]: Alternatively we could have constant strings defined in internal and reference them in both locations, I think that could work and I can address it in a separate PR.

File: sdk/azcore/log.go [hunk 21]
```diff
@@ -31,86 +30,38 @@ const (
...
-func (l *Logger) resetClassifications() { //nolint:unused
-	l.cls = nil
-}
-
-var logger Logger
-
-// Log returns the process-wide logger.
-func Log() *Logger {
-	return &logger
+func SetListener(lst Listener) {
```
Comment (catalinaperalta) [inline sdk/azcore/log.go:56]: Nit: missing doc comment

**Review by catalinaperalta — APPROVED** [review]

File: sdk/azcore/version.go [hunk 22]
```diff
@@ -10,5 +10,5 @@ const (
 	UserAgent = "azcore/" + Version
 
 	// Version is the semantic version (see http://semver.org) of this module.
-	Version = "v0.16.3"
+	Version = "v0.17.0"
```
Comment (catalinaperalta) [inline sdk/azcore/version.go:13]: v0.17.0 is the version I already released. 
```suggestion
	Version = "v0.18.0"
```

**Comment by jhendrixMSFT** [comment]
Fixes https://github.com/Azure/azure-sdk-for-go/issues/15046
